File: fila_c++/fila.cpp
#include<stdio.h> // Biblioteca de entrada/saída padrão.
#include<stdlib.h> // Biblioteca padrão de funções.
#include<iostream> // Biblioteca de entrada/saída de C++.

#define MAX 5 // Define o tamanho máximo da fila.

typedef struct fila // Definição da estrutura da fila.
{
  int vetor[MAX]; // Vetor para armazenar os elementos da fila.
  int max; // Capacidade máxima da fila.
  int inicio; // Índice do início da fila.
  int qtdElementos; // Quantidade de elementos na fila.
} FILA; // Nome da estrutura.

void enfileirar(int aux, FILA*); // Função para enfileirar elementos.
int consultarProximo(FILA); // Função para consultar o próximo elemento da fila.
int tamanhoFila(FILA); // Função para verificar o tamanho da fila.
int desenfileirar(FILA*); // Função para desenfileirar elementos.

int main() // Função principal.
{
   setLocale(LC_ALL, "ptb"); // Configura localidade para português.

  FILA filaDoPao; // Cria uma variável do tipo fila.
  filaDoPao.max = MAX; // Define a capacidade máxima da fila.
  filaDoPao.inicio = 0; // Inicializa o índice do início da fila.
  filaDoPao.qtdElementos = 0; // Inicializa a quantidade de elementos na fila.

  int op, valor; // Declaração de variáveis para operações de fila.
  do // Loop do menu.
  {
    system("cls"); // Limpa a tela.
    printf("\n\n >> FILA << "); // Mostra o título do programa.
    printf("\n\t 1 -> Inserir"); // Opção para inserir elemento.
    printf("\n\t 2 -> Consultar"); // Opção para consultar o próximo elemento.
    printf("\n\t 3 -> Remover"); // Opção para remover elemento.
    printf("\n\t 4 -> Tamanho"); // Opção para consultar o tamanho da fila.
    printf("\n\t 0 -> sair"); // Opção para sair do programa.
    printf("\n\t Digite:  "); // Solicitação de entrada.
    scanf("%d", &op); // Leitura da escolha do usuário.
    switch(op) // Verifica a escolha do usuário.
    {
      case 1: // Caso de inserção.
         printf("\n\n Digite um valor inteiro:  "); // Solicitação de valor.
         scanf("%d", &valor); // Leitura do valor a ser inserido.
         enfileirar(valor, &filaDoPao); // Chama a função para enfileirar o valor.
      break; // Sai do switch.
      case 2: // Caso de consulta.
            printf("\n\n O próximo da FILA é: %d \n\n", consultarProximo(filaDoPao)); // Mostra o próximo elemento.
            system("pause"); // Pausa a execução.
      break; // Sai do switch.
      case 3: // Caso de remoção.
            printf("\n\n Valor REMOVIDO da FILA : %d \n\n", desenfileirar(&filaDoPao)); // Mostra o valor removido.
            system("pause"); // Pausa a execução.
      break; // Sai do switch.
      case 4: // Caso de consulta do tamanho.
            printf("\n\n O tamanho da FILA é: %d \n\n", tamanhoFila(filaDoPao)); // Mostra o tamanho da fila.
            system("pause"); // Pausa a execução.
      break; // Sai do switch.
      case 0: // Caso de sair do programa.
        printf("\n\n ... saindo ... \n\n"); // Mensagem de saída.
        break; // Sai do switch.
      default: // Caso de escolha inválida.
        printf("\n\n .<< Digite uma opcao valida >> \n\n"); // Mensagem de opção inválida.
        system("pause"); // Pausa a execução.
    }
  }
  while(op != 0); // Continua o loop até o usuário escolher sair.

  return 0; // Retorna 0 para indicar sucesso.
};


void enfileirar(int aux, FILA* Fila){ // Função para enfileirar elementos.
  if(Fila->qtdElementos < Fila->max){ // Verifica se a fila não está cheia.
      Fila->vetor[(Fila->inicio + Fila->qtdElementos) % Fila->max] = aux; // Adiciona o elemento na posição correta.
      Fila->qtdElementos += 1; // Incrementa a quantidade de elementos na fila.
  }else{ // Se a fila estiver cheia.
           printf("\n\n <<FILA CHEIA>> \n\n"); // Mensagem de fila cheia.
           system("pause"); // Pausa a execução.
     
  }
 };

 int consultarProximo(FILA Fila) // Função para consultar o próximo elemento da fila.
 {
  return Fila.vetor[Fila.inicio]; // Retorna o próximo elemento da fila.
 };

 int tamanhoFila(FILA Fila) // Função para verificar o tamanho da fila.
 {
  return Fila.qtdElementos; // Retorna a quantidade de elementos na fila.
 };

 int desenfileirar(FILA* Fila) // Função para desenfileirar elementos.
 {
  if (Fila->qtdElementos > 0) // Verifica se a fila não está vazia.
  {
    int aux = Fila->vetor[Fila->inicio]; // Armazena o elemento a ser removido.
    Fila->inicio = (Fila->inicio + 1) % Fila->max; // Atualiza o índice de início.
    Fila->qtdElementos -= 1; // Decrementa a quantidade de elementos na fila.
    return aux; // Retorna o elemento removido.
  }
  else // Se a fila estiver vazia.
  {
    printf("\n\n <<FILA VAZIA>> \n\n"); // Mensagem de fila vazia.
    system("pause"); // Pausa a execução.
    return -1; // Retorna -1 indicando erro.
  }
 };

File: fila_rust/fila.rs
use std::io;

const MAX: usize = 5;

struct Fila {
    vetor: [i32; MAX],
    max: usize,
    inicio: usize,
    qtd_elementos: usize,
}

impl Fila {
    fn new() -> Fila {
        Fila {
            vetor: [0; MAX],
            max: MAX,
            inicio: 0,
            qtd_elementos: 0,
        }
    }

    fn enfileirar(&mut self, aux: i32) {
        if self.qtd_elementos < self.max {
            let pos = (self.inicio + self.qtd_elementos) % self.max;
            self.vetor[pos] = aux;
            self.qtd_elementos += 1;
        } else {
            println!("\n\n <<FILA CHEIA>> \n\n");
        }
    }

    fn consultar_proximo(&self) -> Option<i32> {
        if self.qtd_elementos > 0 {
            Some(self.vetor[self.inicio])
        } else {
            println!("\n\n <<FILA VAZIA>> \n\n");
            None
        }
    }

    fn tamanho_fila(&self) -> usize {
        self.qtd_elementos
    }

    fn desenfileirar(&mut self) -> Option<i32> {
        if self.qtd_elementos > 0 {
            let aux = self.vetor[self.inicio];
            self.inicio = (self.inicio + 1) % self.max;
            self.qtd_elementos -= 1;
            Some(aux)
        } else {
            println!("\n\n <<FILA VAZIA>> \n\n");
            None
        }
    }
}

fn main() {
    let mut fila_do_pao = Fila::new();

    loop {
        println!("\n\n >> FILA << ");
        println!("\t 1 -> Inserir");
        println!("\t 2 -> Consultar");
        println!("\t 3 -> Remover");
        println!("\t 4 -> Tamanho");
        println!("\t 0 -> sair");
        println!("\t Digite:  ");

        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("Erro ao ler entrada");
        let op: i32 = match input.trim().parse() {
            Ok(num) => num,
            Err(_) => continue,
        };

        match op {
            1 => {
                println!("\n\n Digite um valor inteiro:  ");
                let mut valor = String::new();
                io::stdin().read_line(&mut valor).expect("Erro ao ler entrada");
                let valor: i32 = match valor.trim().parse() {
                    Ok(num) => num,
                    Err(_) => continue,
                };
                fila_do_pao.enfileirar(valor);
            }
            2 => {
                if let Some(proximo) = fila_do_pao.consultar_proximo() {
                    println!("\n\n O próximo da FILA é: {} \n\n", proximo);
                }
                io::stdin().read_line(&mut String::new()).unwrap(); // Pause
            }
            3 => {
                if let Some(removido) = fila_do_pao.desenfileirar() {
                    println!("\n\n Valor REMOVIDO da FILA : {} \n\n", removido);
                }
                io::stdin().read_line(&mut String::new()).unwrap(); // Pause
            }
            4 => {
                println!("\n\n O tamanho da FILA é: {} \n\n", fila_do_pao.tamanho_fila());
                io::stdin().read_line(&mut String::new()).unwrap(); // Pause
            }
            0 => {
                println!("\n\n ... saindo ... \n\n");
                break;
            }
            _ => {
                println!("\n\n .<< Digite uma opcao valida >> \n\n");
                io::stdin().read_line(&mut String::new()).unwrap(); // Pause
            }
        }
    }
}
